integration/test/cloudwatchlogs/publish_logs_test.go

```diff
@@ -0,0 +1,96 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+
+//go:build linux && integration
+// +build linux,integration
+
+package cloudwatchlogs
+
+import (
+	"context"
+	"github.com/aws/amazon-cloudwatch-agent/integration/test"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/feature/ec2/imds"
+	"log"
+	"strconv"
+
+	"testing"
+	"time"
+)
+
+const (
+	configOutputPath = "/opt/aws/amazon-cloudwatch-agent/bin/config.json"
+	logScriptPath    = "resources/write_logs.sh"
+	agentRunTime     = 1 * time.Minute
+)
+
+type input struct {
+	testName        string
+	iterations      int
+	numExpectedLogs int
+	configPath      string
+}
+
+var testParameters = []input{
+	{
+		testName:        "Happy path",
+		iterations:      100,
+		numExpectedLogs: 200,
+		configPath:      "resources/config_log.json",
+	},
+	{
+		testName:        "Client-side log filtering",
+		iterations:      100,
+		numExpectedLogs: 100,
+		configPath:      "resources/config_log_filter.json",
+	},
+}
+
+func TestWriteLogsToCloudWatch(t *testing.T) {
+	// this uses the {instance_id} placeholder in the agent configuration,
+	// so we need to determine the host's instance ID for validation
+	ctx := context.Background()
+	c, err := config.LoadDefaultConfig(ctx)
+	if err != nil {
+		// fail fast so we don't continue the test
+		t.Fatalf("Error occurred while creating SDK config: %v", err)
+	}
+	client := imds.NewFromConfig(c)
+	metadata, err := client.GetInstanceIdentityDocument(ctx, &imds.GetInstanceIdentityDocumentInput{})
+	if err != nil {
+		t.Fatalf("Error occurred while retrieving EC2 instance ID: %v", err)
+	}
```

Comment:
Could use `assert`.
```assert.NilError(t, err, "blah %v", err)```

Reply:
I agree, normally, but as I've learned in doing this, doing something like `assert.NoError(t, err)` does not fail fast. So after the test gets an error trying to read from IMDS, it will continue and then panic. In the grand scheme of things, it's still a test so it probably doesn't mean much harm either way, but basically what was happening was I was seeing errors about trying to access/reference nil further down in the tests, which is an immediate red herring to the root of the issue. `t.Fatal()` logs the message and then calls `t.FailNow()`, which pulls the plug on the test. 

That being said, if you still feel strongly about using `assert` in this scenario, I can switch it back. I do realize that me writing code like this is contrary to most of my previous review comments related to these types of assertions 😃

Reply:
I specifically looked that up and thought it would fail fast?!
https://pkg.go.dev/gotest.tools/assert#hdr-Assert_and_Check
>Assert() and Check() both accept a Comparison, and fail the test when the comparison fails. The one difference is that Assert() will end the test execution immediately (using t.FailNow()) whereas Check() will fail the test (using t.Fail()), return the value of the comparison, then proceed with the rest of the test case. 

I am fine with keeping it as is now that I know why.